File: lib/triplet_loss.dart
// file: triplet_loss_gpu.dart
import 'gpu_tensor.dart';
import 'nn.dart';

class TripletLossGPU extends Module {
  final double margin;

  TripletLossGPU({this.margin = 0.2});

  Tensor forward(
    Tensor anchor,
    Tensor positive,
    Tensor negative,
    List<Tensor> tracker,
  ) {
    // 1. Calculate squared Euclidean distances
    final diffPos = anchor - positive;
    final diffNeg = anchor - negative;

    final distP = diffPos.pow(2.0).sum();
    final distN = diffNeg.pow(2.0).sum();

    // 2. Triplet Formula: L = distP - distN + margin
    final rawLoss = (distP - distN) + margin;

    // 3. Hinge Loss: L = max(0, L)
    // This prevents the gradient from pulling embeddings into infinity
    // once the margin is already met.
    final loss = rawLoss.relu();

    tracker.addAll([diffPos, diffNeg, distP, distN, rawLoss, loss]);

    // 🔍 Optional Debugging (keep it minimal for performance)
    if (loss.fetchData()[0] > 100) {
      print("⚠️ High Loss detected: ${loss.fetchData()[0]}");
    }

    return loss;
  }

  @override
  List<Tensor> parameters() => [];
}

File: python/dart_cuda/core/utils/triplet_loss.py
"""Triplet loss — mirrors `lib/core/utils/triplet_loss.dart`."""

from __future__ import annotations

from typing import List

from ..layers.nn import Module
from ..tensor.gpu_tensor import Tensor


class TripletLossGPU(Module):
    def __init__(self, *, margin: float = 0.2):
        self.margin = margin

    def forward(
        self,
        anchor: Tensor,
        positive: Tensor,
        negative: Tensor,
        tracker: List[Tensor],
    ) -> Tensor:
        diff_pos = anchor - positive
        diff_neg = anchor - negative
        dist_p = diff_pos.pow(2.0).sum()
        dist_n = diff_neg.pow(2.0).sum()
        raw_loss = (dist_p - dist_n) + self.margin
        loss = raw_loss.relu()
        tracker.extend([diff_pos, diff_neg, dist_p, dist_n, raw_loss, loss])
        return loss

    def parameters(self) -> List[Tensor]:
        return []
